eio/tls_eio.ml

```diff
@@ -5,198 +5,143 @@ exception Tls_failure of Tls.Engine.failure
-          data
-
-      | Error (alert, `Response resp) ->
-          t.state <- `Error (Tls_failure alert) ;
-          write_t t resp; read_react t
-    in
 
-    match t.state with
-    | `Error e  -> raise e
-    | `Eof      -> raise End_of_file
-    | `Active _ ->
-        let n = read_t t t.recv_buf in
-        match (t.state, n) with
-        | (`Active tls, n) -> handle tls (Cstruct.sub t.recv_buf 0 n)
-        | (`Error e, _)    -> raise e
-        | (`Eof, _)        -> raise End_of_file
-
-  let rec read t buf =
-
-    let writeout res =
-      let open Cstruct in
-      let rlen = length res in
-      let n    = min (length buf) rlen in
-      blit res 0 buf 0 n ;
+  let rec read t flow_buf : int =
+
+    let write_application_data data =
+      let data_len = Cstruct.length data in
+      let n        = min (Cstruct.length flow_buf) data_len in
+      Cstruct.blit data 0 flow_buf 0 n ;
       t.linger <-
-        (if n < rlen then Some (sub res n (rlen - n)) else None) ;
+        if n < data_len
+        then Cstruct.sub data n (data_len -n)
+        else Cstruct.empty ;
       n in
 
-    match t.linger with
-    | Some res -> writeout res
-    | None     ->
-        match read_react t with
-          | None     -> read t buf
-          | Some res -> writeout res
-
-  let writev t css =
-    match t.state with
-    | `Error err  -> raise err
-    | `Eof        -> raise End_of_file
-    | `Active tls ->
-        match Tls.Engine.send_application_data tls css with
-        | Some (tls, tlsdata) ->
-            ( t.state <- `Active tls ; write_t t tlsdata )
-        | None -> invalid_arg "tls: write: socket not ready"
-
-  let write t cs = writev t [cs]
-
-  (*
-   * XXX bad XXX
-   * This is a point that should particularly be protected from concurrent r/w.
-   * Doing this before a `t` is returned is safe; redoing it during rekeying is
-   * not, as the API client already sees the `t` and can mistakenly interleave
-   * writes while this is in progress.
-   * *)
+    if not @@ Cstruct.is_empty t.linger then
+      write_application_data t.linger
+    else
+      match read_t t with
+      | Some data -> write_application_data data
+      | None -> read t flow_buf
+
+  let write t cs =
+    match Tls.Engine.send_application_data t.tls [cs] with
+    | Some (tls, tlsdata) ->
+      t.tls <- tls ;
+      write_t t tlsdata
+    | None -> invalid_arg "tls: write: socket not ready"
+
   let rec drain_handshake t =
-    let push_linger t mcs =
-      match (mcs, t.linger) with
-      | (None, _)         -> ()
-      | (scs, None)       -> t.linger <- scs
-      | (Some cs, Some l) -> t.linger <- Some (Cstruct.append l cs)
-    in
-    match t.state with
-    | `Active tls when not (Tls.Engine.handshake_in_progress tls) ->
-        t
-    | _ ->
-        let cs = read_react t in
-        push_linger t cs; drain_handshake t
+    if not (Tls.Engine.handshake_in_progress t.tls) then t
+    else
+      let application_data = read_t t in
```

Comment:
Again, I don't think the naming is correct here. the buffer returned by `read_t` is some encrypted TLS data, no? and this may be handshake, application data, alerts, heartbeat, ..

Reply:
oops, sorry, `read_t` now indeed returns application data.